source/secure_tunneling.c

```diff
@@ -414,16 +431,42 @@ static int s_secure_tunneling_send(
     if (secure_tunneling_init_send_frame(&frame_options, secure_tunnel, data, type) != AWS_OP_SUCCESS) {
         return AWS_OP_ERR;
     }
-
     return aws_websocket_send_frame(secure_tunnel->websocket, &frame_options);
 }
 
+static bool s_can_send_data_status(void *user_data) {
+    struct aws_secure_tunnel *secure_tunnel = (struct aws_secure_tunnel *)user_data;
+    bool temp = secure_tunnel->can_send_data;
+    return temp;
+}
+
 static int s_secure_tunneling_send_data(struct aws_secure_tunnel *secure_tunnel, const struct aws_byte_cursor *data) {
     if (secure_tunnel->stream_id == INVALID_STREAM_ID) {
         AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Invalid Stream Id");
         return AWS_ERROR_IOTDEVICE_SECUTRE_TUNNELING_INVALID_STREAM;
     }
-    return s_secure_tunneling_send(secure_tunnel, data, DATA);
+    struct aws_byte_cursor new_data = *data;
+    while (new_data.len) {
+        size_t bytes_max = new_data.len;
+        size_t amount_to_send = bytes_max < AWS_IOT_ST_SPLIT_MESSAGE_SIZE ? bytes_max : AWS_IOT_ST_SPLIT_MESSAGE_SIZE;
+
+        struct aws_byte_cursor send_cursor = aws_byte_cursor_advance(&new_data, amount_to_send);
+        if (send_cursor.len) {
+            secure_tunnel->can_send_data = false;
+            if (s_secure_tunneling_send(secure_tunnel, &send_cursor, DATA) != AWS_OP_SUCCESS) {
+                AWS_LOGF_ERROR(AWS_LS_IOTDEVICE_SECURE_TUNNELING, "Failure writing data to out_buf");
+                return AWS_OP_ERR;
+            }
+        }
+        aws_mutex_lock(&secure_tunnel->send_data_mutex);
```

Comment:
I would move the lock-wait-unlock to the top of the loop.  Conceptually it makes sense to check if you can send before sending rather than after (for the next iteration).  You'll also need to move the "secure_tunnel->can_send_data=false;" to inside the lock's protection, after the wait, so it will now be "lock-wait-setfalse-unlock"